pkg/cache/cache.go

```diff
@@ -0,0 +1,283 @@
+// Cache only allows adding objects of type `objectType`
+func NewResourceCache(args ResourceCacheArgs) ResourceCache {
+
+	return &calicoCache{
+		threadSafeCache: cache.New(cache.NoExpiration, cache.DefaultExpiration),
+		workqueue:       workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		calicoClient:    args.Client,
+		synced:          false,
+		ListFunc:        args.ListFunc,
+		KeyFunc:         args.KeyFunc,
+		ObjectType:      args.ObjectType,
+	}
+}
+
+func (c *calicoCache) Set(key string, newObj interface{}) {
+
+	if reflect.TypeOf(newObj).String() == c.ObjectType {
+
+		if existingObj, found := c.threadSafeCache.Get(key); found {
+
+			glog.Debugf("%#v found in cache. comparing..", existingObj)
+			diff := pretty.Compare(existingObj, newObj)
+			glog.Debugf("Diff: %s", diff)
+
+			if len(diff) != 0 {
+
+				glog.Debugf("%#v and %#v do not match.Updating it in calico cache.", newObj, existingObj)
+
+				c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+				c.workqueue.Add(key)
+			}
+		} else {
+			glog.Debugf("%#v not found in calico cache. Adding it in calico cache", newObj)
+
+			c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+			c.workqueue.Add(key)
+		}
+	} else {
+		glog.Errorf("%#v is not of type %s. Not storing in cache.", newObj, c.ObjectType)
+	}
+}
+
+func (c *calicoCache) Delete(key string) {
+
+	glog.Debug("Deleting %s in calico", key)
+	c.threadSafeCache.Delete(key)
+	c.workqueue.Add(key)
+}
+
+func (c *calicoCache) Get(key string) (interface{}, bool) {
+
+	obj, found := c.threadSafeCache.Get(key)
+	if found {
+		return obj, true
+	}
+	return nil, false
+}
+
+// Prime funtion adds object to threadSafeCache.
+// Only difference with Set() call is it does not queue the key
+// to workqueue. Stores pointer of value object in  cache
+func (c *calicoCache) Prime(key string, value interface{}) {
+
+	c.threadSafeCache.Set(key, &value, cache.NoExpiration)
+}
+
+// ListKeys returns list of calico cache keys in the form
+// of array of strings. Cache stores name field of objects as key.
+func (c *calicoCache) ListKeys() []string {
+
+	cacheItems := c.threadSafeCache.Items()
+	keys := make([]string, 0, len(cacheItems))
+	for k := range cacheItems {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+func (c *calicoCache) SetIndexer(indexer k8sCache.Indexer) {
+	c.k8sCacheIndexer = indexer
+}
+
+func (c *calicoCache) GetQueue() workqueue.RateLimitingInterface {
+	return c.workqueue
+}
+func (c *calicoCache) HasSynced() bool {
+	return c.synced
+}
+
+// Load all the calico datastore objects at the begining of run
+func (c *calicoCache) Run(stopChan chan struct{}) {
+
+	// how should we handled failed priming?
+	// with failed priming calico cache will be empty
+	// but it will still function. Only any manual
+	// additions of objects in ETCD datastore will not
+	// be detected and ETCD datastore get bombarded with
+	// events for all k8s objects.
+	c.primeCache()
```

Comment:
Good question - I think the only times it will fail are if we fail to connect to the etcd datastore, right?

Given that, I think this code should probably retry on failure.

e.g.
```go
for err := c.printCache(); err != nil {
  glog.WithError(err).Errorf("Failed to prime Calico cache, retrying")
}
```